### ISSUE TITLE
Implementation of abs does not match specification

### ISSUE THREAD

**Issue by bchetioui** [openxla/stablehlo#1083]
### What happened?

The specification indicates that `abs` on a tensor with complex element type returns a tensor with element type the element type of the complex type (i.e. `c64 -> f32`, `c128 -> f64`).

However, the [reference implementation disagrees with that](https://github.com/openxla/stablehlo/blob/main/stablehlo/dialect/StablehloOps.td#L196-L201), making `AbsOp` a `UnaryElementwiseOp`, where `UnaryElementWiseOp` has the `SameOperandsAndResultType` trait.

The specification is also wrong, in that `abs` is said to be able to return complex types. Given that complex types should map to floating point types, this should not be possible.

### Version information

f9e9d0f2c0dc367c646f90e36d09a040bea9f046

**Comment by sdasgup3** [comment]
>> However, the [reference implementation disagrees with that] .. 

That code is just for custom printing the op. @GleasonK Can talk more about this.

>> The specification is also wrong

Thanks a lot for catching this. That is an oversight. We will update the spec.

**Comment by GleasonK** [comment]
To provide more detail - The `AbsOp` is a `UnaryElementwiseOp` which has a `SameOperandsAndResultShape` trait. The `SameOperandsAndResultType` noted is actually a custom printer which _only_ simplifies printing if the operand and result types match, allowing for printing/parsing of the following syntax:

```
// Input/output different:
%0 = "stablehlo.abs"(%arg0) : (tensor<complex<f32>>) -> tensor<f32>

// Same input and output
%1 = "stablehlo.abs"(%arg0) : tensor<f32>
```

`AbsOp` does support mismatched types without failing verification.

**Comment by GleasonK** [comment]
Regardless - Thanks for pointing out the issue in the spec! Repurposing this ticket to that fix, will be closed by #1001.
